### PR TITLE
[config-plugins] make bundle identifier getter aware of multiple targets

### PR DESCRIPTION
**Pull request by dsokal**
# Why

In EAS CLI, we want to support managed credentials for multitarget iOS projects.
An example of a multitarget project is a project with a Share Extension. Such a project consists of a (main) application target and an extension target. They have distinct bundle identifiers. Currently, there's no way of reading the bundle id for a particular target. We always read the bundle id for the first native target defined in the .pbxproj file.

# How

The bundle identifier for a particular target is stored in an XCBuildConfiguration entry in .pbxproj. The entry can be found by the target name and build configuration pair, e.g. `target = myproject, build configuration = Release`.

- I made `IOSConfig.BundleIdentifier.getBundleIdentifierFromPbxproj` accept a target name and build configuration. Unless the target name is passed the function returns the bundle identifier for the first native target found in the .pbxproj file (this is legacy behavior). 
- `IOSConfig.ProvisioningProfile.setProvisioningProfileForPbxproj` is used in https://github.com/expo/eas-build/blob/main/packages/build-tools/src/ios/configure.ts to configure the Xcode project to use the provisioning profile provided by a user for building the project. I made the function accept the build configuration so only the appropriate XCBuildConfiguration entry is updated. Previously, we would update all build configurations regardless if they were used to build the project.
- I added a function `findApplicationTargetWithDependenciesAsync` that returns a tree of targets for a build scheme. It'll be used to resolve all targets that should have credentials set up.

# Test Plan

- Unit tests.
- Tested in EAS CLI in https://github.com/expo/eas-cli/pull/414

### CODE REVIEW

File: packages/config-plugins/package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
     "fs-extra": "9.0.0",
     "getenv": "^1.0.0",
     "glob": "7.1.6",
+    "nullthrows": "^1.1.1",
```
Comment (EvanBacon) [inline packages/config-plugins/package.json:44]: We use assert in this package

File: packages/config-plugins/src/ios/Target.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+export function getNativeTargets(project: XcodeProject): NativeTargetSectionEntry[] {
+  const section = project.pbxNativeTargetSection();
+  return Object.entries(section).filter(isNotComment);
+}
+
+export function findFirstNativeTarget(project: XcodeProject): NativeTargetSectionEntry {
+  const targets = getNativeTargets(project);
+  return targets[0];
```
Comment (wkozyra95) [inline packages/config-plugins/src/ios/Target.ts:59]: check for undefined here

File: packages/config-plugins/src/ios/ProvisioningProfile.ts [hunk 3]
```diff
@@ -15,27 +14,27 @@ type ProvisioningProfileSettings = {
...
-  } else if (!target) {
-    throw new Error(`Unable to find any targets in this Xcode project.`);
-  }
-
-  const [nativeTargetId, nativeTarget] = target;
-
-  getBuildConfigurationForId(project, nativeTarget.buildConfigurationList)
-    .filter(([, item]: ConfigurationSectionEntry) => item.buildSettings.PRODUCT_NAME)
+  getBuildConfigurationsForListId(project, nativeTarget.buildConfigurationList)
+    .filter(([, item]: ConfigurationSectionEntry) => item.name === buildConfiguration)
```
Comment (wkozyra95) [inline packages/config-plugins/src/ios/ProvisioningProfile.ts:37]: it's used only on builders so it does not matter much, but you are overwriting signing config for all build configurations

File: packages/config-plugins/src/ios/Target.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  const dependencies: Target[] = applicationTarget.dependencies.map(({ value }) => {
+    const { target: targetId } = project.getPBXGroupByKeyAndType(
+      value,
+      'PBXTargetDependency'
+    ) as PBXTargetDependency;
+
+    const [, target] = findNativeTargetById(project, targetId);
+
+    const type =
+      target.productType === TargetType.EXTENSION ||
```
Comment (wkozyra95) [inline packages/config-plugins/src/ios/Target.ts:35]: are you sure that it's only for extensions?

**Review by wkozyra95 — APPROVED** [review]

**Pull request merged**